### PR TITLE
As a guest, I can register - Testing

### PR DESCRIPTION
**Pull request by hoangmirs**
Solve https://github.com/hoangmirs/go-scraper/issues/2

## What happened 👀

- Setup the basic testing
- Add tests for registration feature

## Insight 📝

- As [Ginkgo](https://github.com/onsi/ginkgo) is similar to RSpec, and is also provides some functions that help us to set up and tear down, so I picked it instead of using GoConvey (the default of Beego)
- Modify bootstrap initialization to load the config from `con/app.conf` for test environment. When running the test, it doesn't load the correct config as we didn't call `web.Run()` method, so I need to load it manually.
- Create `RootDir()` method to return the root directory of the project, it is used for getting the absolute path of config file.
- Create the `*_suite_test.go` on each folder to set up for each suite. But I used this command to generate it:
```
cd {folder}
ginkgo bootstrap
```

- Because in forms and controllers, we save data to DB, so I need to initialize the DB when starting the suite and truncate the data after each test. That will ensure that we have a clean DB before running other tests
- Create `test/flash_message.go` helper to get the flash message in tests. The content I copied from Beego's source code
- Modify the test command in Makefile to start and stop the Docker services when testing
- Remove the default test file of Beego

## Proof Of Work 📹

Tests were passed
https://github.com/hoangmirs/go-scraper/actions/runs/485025231

### CODE REVIEW

File: controllers/registration_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+					"password":              {""},
+					"password_confirmation": {""},
+				}
+				body := strings.NewReader(form.Encode())
+				request, _ := http.NewRequest("POST", "/register", body)
+				request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+				response := httptest.NewRecorder()
+				web.BeeApp.Handlers.ServeHTTP(response, request)
+
+				Expect(response.Code).To(Equal(http.StatusOK))
```
Comment (gutakk) [inline controllers/registration_test.go:83]: I think it is weird when we given invalid params but response code is 200. Should this be 400?
What do you think?

File: controllers/registration_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+					"password_confirmation": {""},
+				}
+				body := strings.NewReader(form.Encode())
+				request, _ := http.NewRequest("POST", "/register", body)
+				request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+				response := httptest.NewRecorder()
+				web.BeeApp.Handlers.ServeHTTP(response, request)
+				flashMessage := tests.GetFlash(response.Result().Cookies())
+
+				Expect(flashMessage.Data["error"]).To(Equal("Email Can not be empty"))
```
Comment (gutakk) [inline controllers/registration_test.go:99]: Is it possible to upper case only the first letter of the sentence? and also another test case.

File: forms/registration_test.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/beego/beego/v2/client/orm"
+	"github.com/beego/beego/v2/core/logs"
+	"github.com/hoangmirs/go-scraper/forms"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("RegistrationForm", func() {
+	AfterEach(func() {
```
Comment (gutakk) [inline forms/registration_test.go:13]: Should we move the function to something like `test_helper` or something? as I saw the same function in both controllers and forms test.

File: forms/registration_test.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			err := orm.RunSyncdb("default", true, true)
+			if err != nil {
+				logs.Error(err)
+			}
+		}
+	})
+
+	Describe("CreateUser", func() {
+		Context("given valid attributes", func() {
+			It("returns nil", func() {
```
Comment (gutakk) [inline forms/registration_test.go:28]: Should this be `returns nil error`? If I read the test without reading the function I wouldn't know what return with nil value

File: forms/registration_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+					_ = registrationForm1.CreateUser()
+
+					registrationForm2 := forms.RegistrationForm{
+						Email:                "hoang@nimblehq.co",
+						Password:             "123456",
+						PasswordConfirmation: "123456",
+					}
+
+					err := registrationForm2.CreateUser()
+					Expect(err.Error()).To(Equal("pq: duplicate key value violates unique constraint \"user_email_key\""))
```
Comment (gutakk) [inline forms/registration_test.go:68]: From [this comment](https://github.com/hoangmirs/go-scraper/pull/23#issuecomment-762696657). This error message is not user friendly as this error is a common behaviour for the registration.

File: forms/registration_test.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+						Email:                "hoang@nimblehq.co",
+						Password:             "",
+						PasswordConfirmation: "",
+					}
+					err := registrationForm.CreateUser()
+					Expect(err.Error()).To(Equal("Password Can not be empty"))
+				})
+			})
+
+			Context("given an invalid password", func() {
```
Comment (gutakk) [inline forms/registration_test.go:96]: Should this be something like `given too short password` or something else.
`given an invalid password` is unclear IMO.

File: helpers/text_test.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+var _ = Describe("Text", func() {
+	Describe("ToSnakeCase", func() {
+		Context("given a blank string", func() {
+			It("returns blank string", func() {
+				Expect(helpers.ToSnakeCase("")).To(BeEmpty())
+			})
+		})
+
+		Context("given a normal string", func() {
```
Comment (gutakk) [inline helpers/text_test.go:17]: Should we move the normal case above the edge case? It's more standard to group test by `Happy path before sad path`. cc @olivierobert

File: controllers/registration_test.go (hunk 1 above)
Comment (hoangmirs) [reply controllers/registration_test.go:74]: hmm, I think it would make sense when we build API, but in this case, we just re-render the form, so the request returns code 200.
This should check the template that was rendered, but I didn't find any related functions, so I just check the status instead.

File: forms/registration_test.go (hunk 3 above)
Comment (hoangmirs) [reply forms/registration_test.go:12]: I moved it, thanks for pointing it out

File: tests/db_helper.go [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+	_, err := o.Raw("TRUNCATE TABLE \"user\"").Exec()
+	if err != nil {
+		// If table can't be truncated, rebuild all tables (CAUTION: Star and Message db are lost!)
+		// This is only for absolute startup
+		ResetDB()
+	}
+}
+
+func ResetDB() {
+	// If table can't be truncated, rebuild all tables (CAUTION: Star and Message db are lost!)
```
Comment (olivierobert) [inline tests/db_helper.go:19]: What is this mention about 🙃 

> CAUTION: Star and Message db are lost!

There are no star and message DB 🤔

File: bootstrap/db.go [hunk 9]
```diff
@@ -3,6 +3,8 @@ package bootstrap
 import (
 	"fmt"
 
+	_ "github.com/hoangmirs/go-scraper/models" // Models
```
Comment (olivierobert) [inline bootstrap/db.go:6]: Why are models imported here?

File: controllers/controllers_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+	"github.com/hoangmirs/go-scraper/helpers"
+	_ "github.com/hoangmirs/go-scraper/routers"
+
+	"github.com/beego/beego/v2/server/web"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestControllers(t *testing.T) {
+	RegisterFailHandler(Fail)
```
Comment (olivierobert) [inline controllers/controllers_suite_test.go:16]: If anyone is wondering where `RegisterFailHandler` si coming from: https://onsi.github.io/gomega/ 🧐

File: controllers/registration_test.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+)
+
+var _ = Describe("RegistrationController", func() {
+	AfterEach(func() {
+		tests.ClearUserTable()
+	})
+
+	Describe("GET", func() {
+		It("returns status OK", func() {
+			request, _ := http.NewRequest("GET", "/register", nil)
```
[the remark this answers is not in the document]
Comment (olivierobert) [reply controllers/registration_test.go:24]: @gutakk I added a comment on that closed RFC: https://github.com/nimblehq/compass/issues/469#issuecomment-766715723

File: controllers/registration_test.go [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+			Expect(response.Code).To(Equal(http.StatusOK))
+		})
+
+		It("has body data", func() {
+			request, _ := http.NewRequest("GET", "/register", nil)
+			response := httptest.NewRecorder()
+			web.BeeApp.Handlers.ServeHTTP(response, request)
+
+			Expect(response.Body.Len()).To(BeNumerically(">", 0))
```
Comment (olivierobert) [inline controllers/registration_test.go:36]: Instead of asserting on the body content, would it be possible to assert on which view is rendered?

File: tests/flash_message_helper.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+	for _, cookie := range cookies {
+		if cookie.Name == "BEEGO_FLASH" {
+			return DecodeFlash(cookie)
+		}
+	}
+
+	return nil
+}
+
+func DecodeFlash(cookie *http.Cookie) *web.FlashData {
```
Comment (olivierobert) [inline tests/flash_message_helper.go:21]: Minor thing but since `DecodeFlash` is only used by `GetFlash`, it could be made "private" by not exporting it:

```suggestion
func decodeFlash(cookie *http.Cookie) *web.FlashData {
```

File: main.go [hunk 14]
```diff
@@ -1,11 +1,14 @@
...
 import (
-	_ "github.com/hoangmirs/go-scraper/bootstrap"
+	"github.com/hoangmirs/go-scraper/bootstrap"
+	_ "github.com/hoangmirs/go-scraper/routers" // Routers
 
 	"github.com/beego/beego/v2/server/web"
 )
 
 func main() {
+	bootstrap.SetUp()
```
Comment (olivierobert) [inline main.go:11]: In `bootstrap/bootstrap.go`, why not having `SetUp()` in `init()` so that it would be run when importing it?

**Review by olivierobert — APPROVED** [review]
Looks good to me. 

Just posted one question. Not a blocker but interested to learn why as it can be a way to define a convention when to use `init` or not.

File: bootstrap/bootstrap.go [hunk 15]
```diff
@@ -1,10 +1,23 @@
...
+	"github.com/hoangmirs/go-scraper/helpers"
 	_ "github.com/hoangmirs/go-scraper/routers" // Routers
+
+	"github.com/beego/beego/v2/core/logs"
+	"github.com/beego/beego/v2/server/web"
 )
 
 func init() {
+	configPath := fmt.Sprintf("%s/conf/app.conf", helpers.RootDir())
+	err := web.LoadAppConfig("ini", configPath)
```
Comment (junan) [inline bootstrap/bootstrap.go:15]: Do we need to define `RootDir()` functions? I think just `err := web.LoadAppConfig("ini", "../conf/app.conf")` should work

File: forms/registration_test.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+				user, err := registrationForm.CreateUser()
+
+				Expect(user.Id).NotTo(BeNil())
+				Expect(err).To(BeNil())
+			})
+		})
+
+		Context("given invalid attributes", func() {
+			Context("given a blank email", func() {
+				It("returns the correct error message and does not create new user", func() {
```
Comment (junan) [inline forms/registration_test.go:33]: Should we use `NOT` instead of `not`?

File: main.go (hunk 14 above)
Comment (hoangmirs) [reply main.go:11]: If we put it in `init()`, it would be run every time we call `bootstrap` module. So I move it to the new function `SetUp()` and call it manually as I just wanna setup DB once

File: tests/test_helpers/request_helper.go [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"io"
+	"net/http"
+	"net/http/httptest"
+
+	"github.com/beego/beego/v2/server/web"
+	"github.com/onsi/ginkgo"
+)
+
+// MakeRequest make a HTTP request and return response
```
Comment (junan) [inline tests/test_helpers/request_helper.go:12]:
```suggestion
// MakeRequest makes a HTTP request and returns response
```

File: bootstrap/bootstrap.go (hunk 15 above)
Comment (hoangmirs) [reply bootstrap/bootstrap.go:15]: I tried it before, but it didn't work in `controllers/controllers_suite_test.go`, I wanted to load the config and setup DB, and it could not load the correct config. So I had to create this function

File: helpers/password_test.go [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Password", func() {
+	Describe("EncryptPassword", func() {
+		Context("given a string", func() {
+			It("returns the encrypted string", func() {
+				Expect(helpers.EncryptPassword("hello password")).To(ContainSubstring("$"))
```
Comment (junan) [inline helpers/password_test.go:14]: This is not related to this test but I think the method name `EncryptPassword` should be `HashPassword` instead as the `bcrypt` library does hashing, not encrypting

**Review by gutakk — APPROVED** [review]

File: helpers/password_test.go (hunk 18 above)
Comment (hoangmirs) [reply helpers/password_test.go]: Thanks, I renamed it

**Review by junan — APPROVED** [review]
